scheduler/core/service/service.py

```diff
@@ -68,3 +76,67 @@ def build_scheduler(start: Time = Time("2018-10-01 08:00:00", format='iso', scal
     semesters = frozenset([Semester.find_semester_from_date(start.to_value('datetime')),
                            Semester.find_semester_from_date(end.to_value('datetime'))])
     return Service(start, end, semesters, sites)
+
+
+def calculate_plans_stats(all_plans: List[Plans],
+                          collector: Collector,
+                          selection: Selection) -> Mapping[ProgramID,Tuple[str,float]]:
+
+    all_programs_visits = {}
+    all_programs_length = {}
+    all_programs_scores = {}
+
+    for plans in all_plans:
+        timeloss = 0
+        n_toos = 0
+        plan_conditions = []
+        completion_fraction = {b:0 for b in Band}
+
+        for plan in plans:
+
+            timeloss+=plan.time_left()
+            plan_score = 0
+
+            for visit in plan.visits:
+                obs = collector.get_observation(visit.obs_id)
+                # check if obs is a too
+                if obs.too_type is not None:
+                    n_toos += 1
+                # add to plan_score
+                plan_score += visit.score
+                # add used contrains
+                plan_conditions.append(obs.constraints.conditions)
+                # check completition
+                program = collector.get_program(obs.belongs_to)
+
+                if program.id in all_programs_visits:
+                    all_programs_visits[program.id] += 1
+                    all_programs_scores[program.id] += visit.score
+                else:
+                    # TODO: This asssumes the observations are not spittable
+                    # TODO: and would change in the future.
+                    all_programs_length[program.id] = len(program.observations())
+                    all_programs_visits[program.id] = 0
+                    all_programs_scores[program.id] = visit.score
+
+                if program.band in completion_fraction:
+                    completion_fraction[program.band] += 1
+                else:
+                    raise KeyError('Missing Band in Program!')
+        plans.night_stats =  NightStats(f'{timeloss} min',
+                                        plan_score,
+                                        Conditions.most_restrictive_conditions(plan_conditions),
+                                        n_toos,
+                                        completion_fraction)
+
+    plans_summary = {}
+    for p_id in all_programs_visits.keys():
```

Comment:
Not super important, but remember you iterate over the keys automatically if you do:
```python
for p_id in all_program_visits:
```
The `.keys()` is irrelevant.